sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/SymbolicValues/NullableSymbolicValue.cs

```diff
@@ -0,0 +1,96 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace SonarAnalyzer.Helpers.FlowAnalysis.Common
+{
+    public class NullableSymbolicValue : SymbolicValue
+    {
+        public SymbolicValue WrappedValue { get; }
+
+        public NullableSymbolicValue(SymbolicValue wrappedValue)
+        {
+            this.WrappedValue = wrappedValue;
+        }
+
+        public override IEnumerable<ProgramState> TrySetConstraint(SymbolicValueConstraint constraint,
+            ProgramState currentProgramState)
+        {
+            if (constraint == null)
+            {
+                return new[] { currentProgramState };
+            }
+
+            if (constraint is ObjectConstraint)
+            {
+                var optionalConstraint = constraint == ObjectConstraint.Null
+                    ? NullableValueConstraint.NoValue
+                    : NullableValueConstraint.HasValue;
+
+                return TrySetConstraint(optionalConstraint, currentProgramState);
+            }
+
+            var oldConstraint = currentProgramState.Constraints.GetValueOrDefault(this) as NullableValueConstraint;
+            if (constraint is NullableValueConstraint)
+            {
+                if (oldConstraint == null)
+                {
+                    return new[] { SetConstraint(constraint, currentProgramState) };
+                }
+
+                if (oldConstraint != constraint)
+                {
+                    return Enumerable.Empty<ProgramState>();
+                }
+
+                return new[] { currentProgramState };
+            }
+
+            return TrySetConstraint(NullableValueConstraint.HasValue, currentProgramState)
+                .SelectMany(ps => WrappedValue.TrySetConstraint(constraint, ps));
+        }
+
+        public override IEnumerable<ProgramState> TrySetOppositeConstraint(SymbolicValueConstraint constraint, ProgramState programState)
+        {
+            var negateConstraint = constraint?.OppositeForLogicalNot;
+
+            if (constraint is BoolConstraint)
+            {
+                return TrySetConstraint(negateConstraint, programState)
+                  .Union(TrySetConstraint(NullableValueConstraint.NoValue, programState));
+            }
+
+            return TrySetConstraint(negateConstraint, programState);
+        }
+
+        public override string ToString()
+        {
+            if (base.identifier != null)
+            {
+                return $"NULLABLE_SV_{base.identifier}";
```

Comment:
should we have the `identifier` per SV type? It was awkward for me to see something like this when I debugged the code:
SV_0, NULLABLE_SV_1...

I was expecting SV_0, NULLABLE_SV_0...